[CodeGen] expand-fp: Change frem expansion criterion

frederik-h opened the pull request:
The existing condition for checking whether or not to expand an frem
instruction in expand-fp is not sufficiently precise.
The expansion on other targets than AMDGPU - which is the only intended
user right now - is only prevented due to the interaction with the
MaxLegalFpConvertBitWidth check.  Relying on this is conceptually wrong
and limits the use of the pass for other targets and further expansions
(e.g. merging with the similar ExpandLargeDivRem pass).

Change the expansion criterion to always expand frem of a given type
for targets that use "Expand" as the legalization action for the 
underlying scalar type and use this to exit the pass early for targets 
which do not require any expansions. This requires to change the
frem legalization action for all targets which do not want frem to 
be expanded in this pass from "Expand" to "LibCall".

frederik-h committed:
[AMDGPU] expand-fp: Change frem expansion criterion

The existing condition for checking whether or not to expand an frem
instruction in the pass is not sufficiently precise.  Right now, it is
sufficient to ensure the correct working of the pass. But this is only
true in conjunction with the existing check for the
MaxLegalFpConvertBitWidth value which happens to exit early on targets
on which the frem condition is insufficient.

The correct working of the pass should not rely on this interaction.
The possibility of using the pass for handling further
expansions:(e.g. merging the very similar ExpandLargDivRem into it) is
also limited by this.

This patch changes the pass to expand frem for a target iff the
target's legalization action for the instruction with the scalar type
corresponding to the instruction type is LibCall but the libcall does
not exist. The legalization action for frem in the AMDGPU backend is
adjusted accordingly.

frederik-h committed:
Revert Operation Action for frem to Expand

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into expand-fp-frem-expansion-criterion

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -1000,8 +1008,8 @@ static bool runImpl(Function &F, const TargetLowering &TLI,
   if (ExpandFpConvertBits != llvm::IntegerType::MAX_INT_BITS)
     MaxLegalFpConvertBitWidth = ExpandFpConvertBits;
 
-  if (MaxLegalFpConvertBitWidth >= llvm::IntegerType::MAX_INT_BITS)
```
We wouldn't have to solve this problem if we could demote frem to an intrinsic. 

Given the pass has to explicitly handle certain types, you might as well just handle those ones. It gets uglier when you start to consider vectors.

frederik-h committed:
[AMDGPU] expand-fp: Add early exit for targets that don't require any expansion

This commit adds a function to check if a target needs expansion
of any type of frem instructions. In conjunction with the trivial
check if the large fp conversion expansions are necessary, this
can be used to perform any early exit from the pass if no
expansions are needed for a target.

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into expand-fp-frem-expansion-criterion

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -74,11 +74,63 @@ class FRemExpander {
+  /// The frem argument/return types that can be expanded by this class.
+  // TODO The expansion could work for other floating point types
+  // as well, but this would require additional testing.
+  inline static const SmallVector<MVT, 3> ExpandableTypes{MVT::f16, MVT::f32,
+                                                          MVT::f64};
```
```suggestion
 static constexpr std::array<MVT, 3> ExpandableTypes{MVT::f16, MVT::f32,
                                                          MVT::f64};
```

arsenm replied:
Or just regular c array

frederik-h replied:
> Or just regular c array

I changed it to the "constexpr std::array" variant.

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -74,11 +74,63 @@ class FRemExpander {
+
+  /// Libcalls for frem instructions of the type at the corresponding
+  /// positions of ExpandableTypes.
+  inline static const SmallVector<RTLIB::Libcall, 3> FremLibcalls{
```
ditto

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -74,11 +74,63 @@ class FRemExpander {
+  /// non-existing libcall.
+  static bool shouldExpandFremType(const TargetLowering &TLI, EVT VT) {
+    TargetLowering::LegalizeAction LA = TLI.getOperationAction(ISD::FREM, VT);
+    if (LA != TargetLowering::LegalizeAction::LibCall &&
+        LA != TargetLowering::LegalizeAction::Expand)
+      return false;
```
This should only handle the Expand case, and shouldn't need to consider Libcall. Libcall is what the DAG will emit. If the actions are still specifying expand, they need to change to report libcall

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -74,11 +74,63 @@ class FRemExpander {
+  /// should happen if the legalization for the scalar type uses a
+  /// non-existing libcall.
+  static bool shouldExpandFremType(const TargetLowering &TLI, EVT VT) {
+    TargetLowering::LegalizeAction LA = TLI.getOperationAction(ISD::FREM, VT);
```
This probably mishandles vectors, but I doubt any target does anything other than scalarize this

frederik-h replied:
I have added an `assert(!VC.isVector())` to document this restriction.

frederik-h committed:
Review changes

* Use constexpr std::array instead of inline const SmallVector.
* shouldExpandFremType
  - Remove LibCall action handling
  - Add assert to document that vectors are not handled
  - additionally: Inline variable

frederik-h committed:
fixup! Review changes

frederik-h committed:
Try fix Windows build problem

The compiler in the Windows build failed to deduced the type in the auto.

frederik-h committed:
Furhter fixup for Windows build

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -74,11 +74,64 @@ class FRemExpander {
+    if (LA != TargetLowering::LegalizeAction::Expand)
+      return false;
+
+    auto Libcall = getFremLibcallForType(VT);
+    return Libcall.has_value() && !TLI.getLibcallName(*Libcall);
```
> That is, you suggest that I change the legalization action for ISD::FREM for every target that currently uses Expand to LibCall (unless it provides a custom lowering function, I suppose; in this case Custom, as e.g. used by the NVPTX target, would be correct?)

Yes. However, for the PTX case, that is the same bug. The custom expansion is the same buggy expansion you're fixing. As a hack until that's also converted, you could just ignore custom too

frederik-h replied:
I have implemented the change now. I think this should go into a separate PR if we want to keep it this way as it affects many targets. I also think it is a bit confusing to make this "Expand" -> "LibCall" change for `ISD::FREM` only. Several times, I had to exclude it from a list of instructions which should probably all be handled in the same way.

frederik-h committed:
Change ISD::FREM legalization actions from Expand to LibCall for scalar types

frederik-h committed:
expand-fp: always expand frem if legalization action is "Expand"

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into expand-fp-frem-expansion-criterion

frederik-h committed:
Revert deletion of comment line

frederik-h committed:
Add back deleted line

frederik-h committed:
clang-format changes

jmmartinez commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -101,7 +101,7 @@ class FRemExpander {
     assert(VT.isSimple() && "Can expand only simple types");
 
-    return find(ExpandableTypes, VT.getSimpleVT());
+    return find(ExpandableTypes, VT.getSimpleVT()) != ExpandableTypes.end();
```
Is this the same as `llvm::is_contained(ExpandableTypes, VT.getSimpleVT())` ?

frederik-h committed:
Replace two function uses by better llvm alternatives

frederik-h committed:
trigger CI

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into expand-fp-frem-expansion-criterion

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into expand-fp-frem-expansion-criterion

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into expand-fp-frem-expansion-criterion

frederik-h committed:
fixup! Merge remote-tracking branch 'upstream/main' into expand-fp-frem-expansion-criterion

frederik-h committed:
trigger build

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into expand-fp-frem-expansion-criterion

frederik-h committed:
Trigger CI

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -74,11 +74,54 @@ class FRemExpander {
+  static bool shouldExpandFremType(const TargetLowering &TLI, EVT VT) {
+    assert(!VT.isVector() && "Cannot handle vector type; must scalarize first");
+    return (TLI.getOperationAction(ISD::FREM, VT) ==
+            TargetLowering::LegalizeAction::Expand);
```
```suggestion
    return TLI.getOperationAction(ISD::FREM, VT) ==
            TargetLowering::LegalizeAction::Expand;
```

arsenm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -534,9 +534,9 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
-  setOperationAction(ISD::FREM, MVT::f80, Expand);
+  setOperationAction(ISD::FREM, MVT::f32, LibCall);
+  setOperationAction(ISD::FREM, MVT::f64, LibCall);
+  setOperationAction(ISD::FREM, MVT::f80, LibCall);
```
Why was this setting an f80 action in the first place?

frederik-h replied:
I don't know. Perhaps AArch64 maintainers can help to answer this? @aemerson, @davemgreen?

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4809,7 +4809,8 @@ void DAGTypeLegalizer::WidenVectorResult(SDNode *N, unsigned ResNo) {
     if (!TLI.isOperationLegalOrCustomOrPromote(N->getOpcode(), WideVecVT) &&
-        TLI.isOperationExpand(N->getOpcode(), VT.getScalarType())) {
+        (TLI.isOperationExpand(N->getOpcode(), VT.getScalarType()) ||
+         TLI.isOperationLibCall(N->getOpcode(), VT.getScalarType()))) {
```
Add isOperationExpandOrLibCall?

Beyond the scope of this patch, but the vector legalization logic is a pain. Ideally this would be written to do vector splitting based on the available calls (which theoretically should include a vector version)

arsenm commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -74,11 +74,54 @@ class FRemExpander {
+  };
+
 public:
   static bool canExpandType(Type *Ty) {
```
This seems redundant with getFremLibcallForType

frederik-h replied:
`getFremLibcallforType` wasn't used anymore and has been removed now.

frederik-h committed:
Apply suggestions from code review

Co-authored-by: Matt Arsenault <arsenm2@gmail.com>

frederik-h committed:
fixup! Apply suggestions from code review

clang-format changes

frederik-h committed:
Review changes

- Remove some dead code from ExpandFp.cpp
- Add and use isOperationExpandOrLibCall
- Fix comment

frederik-h commented on llvm/lib/CodeGen/ExpandFp.cpp:
```diff
@@ -74,11 +74,54 @@ class FRemExpander {
+  /// Libcalls for frem instructions of the type at the corresponding
+  /// positions of ExpandableTypes.
+  static constexpr std::array<RTLIB::Libcall, 3> FremLibcalls{
+      RTLIB::REM_F32, RTLIB::REM_F32, RTLIB::REM_F64};
```
This wasn't used anymore and has been removed now.

frederik-h committed:
fixup! Review changes

clang-format changes

davemgreen reviewed (commented):
I'm not sure about the costs but nothing is great for the cost of a libcall (I thought they were usually 10). The Arm/AArch64 changes look OK to me.

dschuff reviewed (commented):
WebAssembly LGTM

frederik-h committed:
Merge remote-tracking branch 'upstream/main' into expand-fp-frem-expansion-criterion

frederik-h commented:
@zixu-w @SundeepKushwaha @SixWeining @mandlebug @asb @andreisfr @koachan @kaz7 @RKSimon @uweigand Ping. Do you have any concerns about the changes to the legalization actions for the targets you maintain? The changes to the other targets are similar to those approved by davemgreen and dschuff. I do not necessarily need a review of the expand-fp changes - but I am glad about any feedback, of course - as @arsenm has already reviewed this part of the PR.

aankit-ca commented on llvm/lib/Target/Hexagon/HexagonISelLowering.cpp:
```diff
@@ -1661,13 +1661,14 @@ HexagonTargetLowering::HexagonTargetLowering(const TargetMachine &TM,
     for (MVT VT : MVT::integer_valuetypes())
       setOperationAction(IntExpOp, VT, Expand);
   }
+  for (MVT VT : MVT::fp_valuetypes())
```
This seems like a redundant loop

aankit-ca reviewed (commented):
FYI @iajbar

frederik-h committed:
Hexagon: Remove redundant loop

aankit-ca reviewed (commented):
@frederik-h Can you give us some time to evaluate the impact on Hexagon backend?
